src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java

```diff
@@ -166,13 +211,43 @@ private Response createSession(Command command) throws IOException {
         setCommandCodec(new AppiumW3CHttpCommandCodec());
         refreshAdditionalCommands();
         setResponseCodec(dialect.getResponseCodec());
-        return result.createResponse();
+        Response response = result.createResponse();
+
+        setDirectConnect(response);
+
+        return response;
     }
 
     public void refreshAdditionalCommands() {
         getAdditionalCommands().forEach(this::defineCommand);
     }
 
+    @SuppressWarnings("unchecked")
+    private void setDirectConnect(Response response) throws MalformedURLException {
+        if (!this.appiumClientConfig.isDirectConnectEnabled()) {
+            return;
+        }
+
+        Map<String, ?> responseValue = (Map<String, ?>) response.getValue();
+
+        String directConnectProtocol = getDirectConnectValue(responseValue, "directConnectProtocol");
+        String directConnectPath = getDirectConnectValue(responseValue, "directConnectPath");
+        String directConnectHost = getDirectConnectValue(responseValue, "directConnectHost");
+        String directConnectPort = getDirectConnectValue(responseValue, "directConnectPort");
+
+        if (directConnectProtocol == null || directConnectHost == null
+                || directConnectPath == null || directConnectPort == null) {
+            return;
+        }
+        setClientWithURL(new URL(directConnectProtocol + "://" + directConnectHost + ":"
+                + directConnectPort + directConnectPath));
+    }
+
+    private String getDirectConnectValue(Map<String, ?> responseValue, String key) {
+        String directConnectPath = String.valueOf(responseValue.get("appium:" + key));
```

Comment:
can we use a constant instead?

Reply:
also, do we only allow `appium` prefix? I thought it could be any vendor prefix

Reply:
Maybe possible. I just followed https://github.com/webdriverio/webdriverio/blob/6bbe1cf371f06e517ac62a4e02724c8b610803cb/packages/webdriver/src/utils.ts#L292

Afaik, ordinally these caps did not have any vendor prefix.